File: .scripts/Metals/MetalInfo.ts
export type AllomancyGroup = 'physical' | 'mental' | 'enhancement' | 'temporal';
export type FeruchemyGroup = 'physical' | 'cognitive' | 'spiritual' | 'hybrid';
export type Axis = 'internal' | 'external';
export type Polarity = 'pushing' | 'pulling';

export class MetalInfo {
    public name: string;
    public description: string;
    public defName?: string;
    public color: [number, number, number];
    public colorTwo?: [number, number, number];
    public godMetal: boolean = false;
    public stackable: boolean = true;
    public drawSize: number = 1;
    public beauty: number;
    public marketValue: number;
    public maxAmount: number = 100;
    public genesToGrant: Record<string, string[]> = {};
    public allomancy?: AllomancyInfo
    public feruchemy?: FeruchemyInfo;
    public buildable?: BuildableInfo;
    public mining?: MiningInfo;
    public alloy?: AlloyInfo;

    constructor(self: Partial<MetalInfo>) {
        Object.assign(this, self);
    }
}

export class AllomancyInfo {
    public userName?: string;
    public description: string;
    public group: AllomancyGroup | 'None' = 'None';
    public axis: Axis | 'None' = 'None';
    public polarity: Polarity | 'None' = 'None';
    public abilities: string[] = [];

    constructor(self: Partial<AllomancyInfo>) {
        Object.assign(this, self);
    }
}

export class FeruchemyInfo {
    public userName?: string;
    public description: string;
    public group: FeruchemyGroup | 'None' = 'None';
    public abilities: string[] = [];
    public attribute?: string;
    public customHediffClass: boolean = false;
    public store?: FeruchemyAbilityInfo;
    public tap?: FeruchemyAbilityInfo;

    constructor(self: Partial<FeruchemyInfo>) {
        Object.assign(this, self);
        this.store = self.store ? new FeruchemyAbilityInfo(self.store) : undefined;
        this.tap = self.tap ? new FeruchemyAbilityInfo(self.tap) : undefined;
    }
}

export class FeruchemyAbilityInfo {
    public description: string;
    public stages: number = 20;
    public multiplyBySeverity: boolean = false;
    public capacityMods: Record<string, { factor?: number; offset?: number }>;
    public statOffsets?: Record<string, number>;
    public statFactors?: Record<string, number>;
    public hungerRateFactor?: number;

    constructor(self: Partial<FeruchemyAbilityInfo>) {
        Object.assign(this, self);
    }
}

export class BuildableInfo {
    public commonality: number;
    public defense?: {
        sharp?: number;
        blunt?: number;
        heat?: number;
        coldInsulation?: number;
        heatInsulation?: number;
    };
    public offense?: {
        sharp?: number;
        blunt?: number;
        armorPenetration?: number;
        cooldown?: number;
    }
    public stuffStatFactors?: Record<string, number>;

    constructor(self: Partial<BuildableInfo>) {
        Object.assign(this, self);
    }
}

export class MiningInfo {
    public description?: string;
    public hitPoints: number;
    public yield: number;
    public commonality: number;
    public sizeRange: [number, number];

    constructor(self: Partial<MiningInfo>) {
        Object.assign(this, self);
    }
}

interface AlloyInput {
    type: 'simple' | 'complex' | 'god';
    ingredients: {
        item: string | string[];
        count: number;
    }[];
    stuff?: string | string[];
    stuffCount?: number;
    product: {
        item?: string;
        count: number;
    };
}

export class AlloyInfo {
    public ingredients: AlloyIngredient[];
    public stuff?: string[];
    public stuffCount?: number;
    public product: AlloyProduct;
    public type: 'simple' | 'complex' | 'god';

    public constructor({type, ingredients, stuff, stuffCount, product = {count: 10}}: AlloyInput) {
        this.type = type;
        this.ingredients = ingredients.map(i => new AlloyIngredient({
            items: i.item ? (Array.isArray(i.item) ? i.item : [i.item]) : undefined,
            count: i.count,
        }));
        this.product = new AlloyProduct({
            item: product.item,
            count: product.count,
        });
        this.stuff = stuff ? (Array.isArray(stuff) ? stuff : [stuff]) : undefined;
        this.stuffCount = stuffCount;
    }
}

export class AlloyIngredient {
    public items?: string[];
    public stuffs?: string[];
    public count: number;

    constructor(self: Partial<AlloyIngredient>) {
        Object.assign(this, self);
    }
}

export class AlloyProduct {
    public item: string;
    public count: number;

    constructor(self: Partial<AlloyProduct>) {
        Object.assign(this, self);
    }
}

File: tools/Cosmere.Tools/Models/MetalInfo.cs
using System.Collections.Generic;
using Newtonsoft.Json;

namespace Cosmere.Tools.Models;

public enum AllomancyGroup {
    None,
    Physical,
    Mental,
    Enhancement,
    Temporal,
}

public enum FeruchemyGroup {
    None,
    Physical,
    Cognitive,
    Spiritual,
    Hybrid,
}

public enum Axis {
    None,
    Internal,
    External,
}

public enum Polarity {
    None,
    Pushing,
    Pulling,
}

public class MetalInfo {
    public string Name { get; set; } = string.Empty;

    public string Description { get; set; } = string.Empty;

    public string? DefName { get; set; }

    public ColorInfo Color { get; set; } = new();

    public ColorInfo? ColorTwo { get; set; }

    public bool GodMetal { get; set; }

    /// <summary>The Shards this metal is made of, with what burning it grants to each.</summary>
    public List<ShardGrantInfo>? Shards { get; set; }

    public bool Stackable { get; set; } = true;

    public float DrawSize { get; set; } = 1f;

    public float Beauty { get; set; }

    public float MarketValue { get; set; }

    public int MaxAmount { get; set; } = 100;

    public Dictionary<string, string[]> GenesToGrant { get; set; } = new();

    public AllomancyInfo? Allomancy { get; set; }

    public FeruchemyInfo? Feruchemy { get; set; }

    public BuildableInfo? Buildable { get; set; }

    public MiningInfo? Mining { get; set; }

    public AlloyInfo? Alloy { get; set; }
}

public class AllomancyInfo {
    public string? UserName { get; set; }

    public string Description { get; set; } = string.Empty;

    public AllomancyGroup Group { get; set; } = AllomancyGroup.None;

    public Axis Axis { get; set; } = Axis.None;

    public Polarity Polarity { get; set; } = Polarity.None;

    public List<string> Abilities { get; set; } = new();
}

public class FeruchemyInfo {
    public string? UserName { get; set; }

    public string Description { get; set; } = string.Empty;

    public FeruchemyGroup Group { get; set; } = FeruchemyGroup.None;

    public List<string> Abilities { get; set; } = new();

    public string? Attribute { get; set; }

    public bool CustomHediffClass { get; set; }

    public FeruchemyAbilityInfo? Store { get; set; }

    public FeruchemyAbilityInfo? Tap { get; set; }

    public float? StoreRateMultiplier { get; set; }

    public float? TapRateMultiplier { get; set; }
}

public class FeruchemyAbilityInfo {
    public string Description { get; set; } = string.Empty;

    public int Stages { get; set; } = 20;

    public bool MultiplyBySeverity { get; set; }

    public Dictionary<string, CapacityMod> CapacityMods { get; set; } = new();

    public Dictionary<string, float>? StatOffsets { get; set; }

    public Dictionary<string, float>? StatFactors { get; set; }

    public float? HungerRateFactor { get; set; }
}

public class CapacityMod {
    public float? Factor { get; set; }

    public float? Offset { get; set; }
}

public class BuildableInfo {
    public float Commonality { get; set; }

    public DefenseInfo? Defense { get; set; }

    public OffenseInfo? Offense { get; set; }

    public Dictionary<string, float>? StuffStatFactors { get; set; }
}

public class DefenseInfo {
    public float? Sharp { get; set; }

    public float? Blunt { get; set; }

    public float? Heat { get; set; }

    public float? ColdInsulation { get; set; }

    public float? HeatInsulation { get; set; }
}

public class OffenseInfo {
    public float? Sharp { get; set; }

    public float? Blunt { get; set; }

    public float? ArmorPenetration { get; set; }

    public float? Cooldown { get; set; }
}

public class MiningInfo {
    public string? Description { get; set; }

    public int HitPoints { get; set; }

    public int Yield { get; set; }

    public float Commonality { get; set; }

    public int[] SizeRange { get; set; } = new int[2];
}

public class AlloyInfo {
    public List<AlloyIngredient> Ingredients { get; set; } = new();

    [JsonConverter(typeof(StringOrListConverter))]
    public List<string>? Stuff { get; set; }

    public int? StuffCount { get; set; }

    public AlloyProduct Product { get; set; } = new();

    public string Type { get; set; } = "simple";
}

public class AlloyIngredient {
    public List<string>? Items { get; set; }

    public List<string>? Stuffs { get; set; }

    public int Count { get; set; }
}

public class AlloyProduct {
    public string Item { get; set; } = string.Empty;

    public int Count { get; set; }
}

/// <summary>One Shard a metal is made of, and the Connection burning it grants there.</summary>
public class ShardGrantInfo {
    public string Shard { get; set; } = string.Empty;

    public int Grant { get; set; }
}
